libs/permit-utils/src/utils/getTokenVersion.ts

```diff
@@ -0,0 +1,32 @@
+import type { JsonRpcProvider } from '@ethersproject/providers'
+
+import { defaultAbiCoder } from '@ethersproject/abi'
+
+import { getContract } from './getContract'
+
+import { VERSION_ABIS } from '../abi/versionAbis'
+
+export async function getVersion(tokenAddress: string, provider: JsonRpcProvider, index = 0): Promise<string> {
+  const abi = VERSION_ABIS[index]
+
+  if (!abi) {
+    return '1'
+  }
+
+  try {
+    const contract = getContract(tokenAddress, abi, provider)
+    const data = contract.interface.encodeFunctionData(abi[0].name)
+    const response = await provider.call({ to: tokenAddress, data })
+
+    if (response === '0x' || Number.isNaN(Number(response))) {
+      return '1'
+    }
+    if (!response.startsWith('0x')) {
+      return response
+    }
+
+    return defaultAbiCoder.decode(['string'], response).toString()
+  } catch (e) {
+    return getVersion(tokenAddress, provider, index + 1)
```

Comment:
should there be a maximum number of attempts? 

What happens if we get a systematic error for other reasons, would we enter an infinite loop?

Reply:
There is!
It stops after the second.
It iterates over the version ABIs, of which we have only 2.
This logic isn't might, though. Except for the abi decoding, it's the same as 1inch's permit utils lib.

Reply:
Ahh, I see, the index is just used in the ABIs... disregard my comment then